### PR TITLE
Implement forward_all function that performs forwarding on mulptiple variables at once

### PR DESCRIPTION
**Pull request by takuseno**
Hi, @TE-TakuyaNarihira san.

I added new function `forward_all`.

Formerly, forwarding multiple outputs which shared hidden layers (just like policy and value of actor-critic) requires multiple forwardings in static graph. See the example below.
```py
x = nn.Variable((3, 3))
h = PF.affine(x, 3, name='hidden')
y1 = PF.affine(h, 3, name='y1')
y2 = PF.affine(h, 3, name='y2')

y1.forward()
y2.forward() # hidden layer was computed twice!!
```
This is critical at huge architectures.

Thus, I added `forward_all` function.
```py
x = nn.Variable((3, 3))
h = PF.affine(x, 3, name='hidden')
y1 = PF.affine(h, 3, name='y1')
y2 = PF.affine(h, 3, name='y2')

# compute h only once!
nn.forward_all([y1, y2])
```
This function performs forwarding with shared fclosed, which prevents shared layers from being revisited.

What do you think of this?

### CODE REVIEW

**Comment by takuseno** [comment]
@TE-AkioHayakawa Hi, Hayakawa san, it's been a week!
Thank you for reviewing my PR! I will apply all of your suggestions in a week!

**Comment by takuseno** [comment]
@TE-AkioHayakawa Hi, Hayakawa san!
I've implemented the unit test! All tests have passed on my machine!

**Comment by TE-AkioHayakawa** [comment]
Sorry for late reply.
I will review your code in about a week.
 
Thank you for your commitment!

**Comment by takuseno** [comment]
Thank you for reviewing my code! I'll check unit tests with Python 2.

**Comment by TE-AkioHayakawa** [comment]
Sorry, I might be wrong. This problem is not related to python 2.

I found that the result gradients are not correct when CpuArray is used (not CpuCachedArray) and backward() is called with clear_buffer=True.

File: python/test/test_forward_all.py [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+    l2 = F.softmax_cross_entropy(z2, t, 1)
+    L2 = F.mean(l2)
+    nn.forward_all([L1, L2])
+
+    # Backprop for z1
+    # Diff should be initialized since they are always accumulated
+    x.g = 0
+    w1.g = 0
+    b1.g = 0
+    L1.backward(clear_buffer=True)
```
Comment (takuseno) [inline python/test/test_forward_all.py]: @TE-AkioHayakawa 
The test has passed with the latest changes. But, I don't know why this `clear_buffer=True` makes this test passed.

**Comment by TE-AkioHayakawa** [comment]
CI test has passed.

**Comment by TE-AkioHayakawa** [comment]
@takuseno I will merge your pull request as soon as possible.
Thank you again for all your contributions!!

**Comment by takuseno** [comment]
@TE-AkioHayakawa Thank you!!
